domain/controllernode/state/state.go

```diff
@@ -350,6 +389,87 @@ func encodeAPIAddresses(controllerID string, addrs []controllernode.APIAddress)
 	return result
 }
 
+// calculateAddressDeltas returns the list of addresses to add, remove, and
+// update from the controller node table given the existing and new addresses.
+// The updated addresses are the list of addresses for which the IsAgent flag
+// has changed.
+func calculateAddressDeltas(existing, new []controllerAPIAddress) (toAdd []controllerAPIAddress, toRemove []controllerAPIAddress, toUpdate []controllerAPIAddress) {
+	existingMap := make(map[string]controllerAPIAddress)
+	newMap := make(map[string]controllerAPIAddress)
+
+	for _, addr := range existing {
+		existingMap[addr.Address] = addr
+	}
+	for _, addr := range new {
+		newMap[addr.Address] = addr
+	}
+
+	for key, addr := range newMap {
+		_, found := existingMap[key]
+		if !found {
+			toAdd = append(toAdd, addr)
+		} else if existingMap[key].IsAgent != addr.IsAgent {
+			toUpdate = append(toUpdate, addr)
+		}
+	}
+	for key, addr := range existingMap {
+		_, found := newMap[key]
+		if !found {
+			toRemove = append(toRemove, addr)
+		} else if existingMap[key].IsAgent != addr.IsAgent {
+			toUpdate = append(toUpdate, addr)
+		}
+	}
+
+	return toAdd, toRemove, toUpdate
+}
+
+// GetAPIAddresses returns the list of API addresses for the provided controller
+// node.
+func (st *State) GetAPIAddresses(ctx context.Context, ctrlID string) ([]controllernode.APIAddress, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	ident := controllerID{ID: ctrlID}
+
+	stmt, err := st.Prepare(`
+SELECT &controllerAPIAddress.* 
+FROM controller_api_address
+WHERE controller_id = $controllerID.controller_id
+`, controllerAPIAddress{}, ident)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var result []controllerAPIAddress
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, ident).GetAll(&result)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return controllernodeerrors.EmptyAPIAddresses
+		} else if err != nil {
+			return errors.Errorf("getting api addresses for controller node %q: %w", ctrlID, err)
+		}
+		return nil
+	}); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return st.decodeAPIAddresses(result), nil
+}
+
+func (st *State) decodeAPIAddresses(addrs []controllerAPIAddress) []controllernode.APIAddress {
+	result := make([]controllernode.APIAddress, 0, len(addrs))
+	for _, addr := range addrs {
+		result = append(result, controllernode.APIAddress{
+			Address: addr.Address,
+			IsAgent: addr.IsAgent,
+		})
+	}
+	return result
+}
+
```

Comment:
This should probably be moved to a `types.go` or `params.go` file

Also, it doesn't need to be a method, make it just a func

Reply:
Done, moved to a func. Left it in state.go because that's where we usually put the encode*/decode* funcs.